Agent-Specific Discovery

woutermont opened the pull request.

woutermont commented on proposals/specification/agent-specific-discovery/index.bs:
```diff
@@ -0,0 +1,438 @@
+
+<h2 class="no-num">Appendix A: Summary Sequence Diagram</h2>
+
+<img class="sequence-diagram" src="diagrams/agent-specific-discovery.mmd.svg" />
```
```suggestion
<img class="sequence-diagram" src="diagrams/agent-specific-discovery.seq.mmd.svg" />
```

woutermont commented:
@tomhgmns

justinwb commented:
@woutermont do you plan to adjust the relevant places in the interop spec as part of this proposal?

woutermont commented:
@justinwb yes, working on it now; will hopefully finish tomorrow!

woutermont commented:
Okay, so I've updated the proposal, taking more terminology from the main document, and then subtituted references to the proposal for some portions of the relevant sections in the main document. I tried to be conservative, so there is still quite a lot of text repeating what is already stated in the proposal.

In trying to align the two as good as possible, I wondered why we want to distinguish Social Agents from Applications at this level. Does the Authorization Agent need to know what type it is, given that they really contain the same info. Especially the double Registry->Registration predicate (`hasApplicationRegistration`/`hasSocialAgentRegistration`) is hard to incorporate. I feel that it only adds redundant data, since we get the same info from the resource it points to: the following paths hold the same information.

```
:registry interop:hasApplicationRegistration :registration .
:registry interop:hasRegistration / rdf:type interop:ApplicationRegistration .
:registry interop:hasRegistration / inter:registeredAgent / rdf:type interop:Application .
```

On another note: the resource hierachy included at the end of the section on Agent Registrations seems to give the wrong hint that it always has to be a filesystem-like structure. I think just including Turtle examples, or graphs like the ones in the Primer, are clearer.

elf-pavlik commented on proposals/specification/agent-management/ard-request.bs:
```diff
@@ -0,0 +1,24 @@
+Once the [=Agent Manager=] of an [=Entity=] is known, an [=Agent=] wishing to
+discover its [=Agent-Specific Resources=] related to that [=Entity=] [SHOULD]
+query the [=Agent Manager=] by sending a [:HEAD:] or [:GET:] request to the
+[=Agent Manager's=] URL in which the value of the [:Authorization:] header is
+set to a [[DPOP]]-bound [[OIDC]] [=Identity Token=] containing an [=Identifier=]
+of the [=Agent=] as value of the `azp` [=Claim=] or as *sole* value of the
+`aud` [=Claim=].
```
If we want to be that specific we should probably reference Solid OIDC.
The ID Token is only pushed as a claim and access token should be used in Authorization header.

woutermont replied:
You're right! This should be an Access Token with the `client_id` set to the URI of the Agent.

elf-pavlik replied:
Solid OIDC doesn't put any more specific requirement on the format of Access Tokens, they are left up to RS & AS.
I think we should use some abstract language and have non-normative reference to Solid-OIDC as a concrete example.

elf-pavlik replied:
@csarven we think this should be discussed on Wednesday

woutermont replied:
ACTION: remove token details, talk about "authorized request" and refer to (yet to be created) solid ecosystem options

elf-pavlik commented on proposals/specification/agent-management/ard-response.bs:
```diff
@@ -0,0 +1,62 @@
+
+If an [:Authorization:] header is present, with the scheme set to [:DPoP:], and
+the included [=Identity Token=] is valid, and it authenticates an [=Agent=] known to
+the [=Agent Manager=], the latter [MUST] respond with either of the following; and it 
```
Why do we need those 3 different options? I think `204` is the safest given that it also can act as Identity Document of the Authorization Agent / Agent Manager

woutermont replied:
The `200` allows for optimization, in which the Agent Manager can already respond with the Agent-Specific Resource in the body.

I added the `303` because to me that is the most correct response when used with `pim:preferencesFile` or `rdfs:seeAlso`. In fact, I would drop the other two status codes and prefer the `pim:preferencesFile` predicate over `interop:hasAgentManager`

woutermont replied:
Agreed in meeting to keep the options.

elf-pavlik reviewed (changes requested):
I'm not sure about specifying registering agents (§5), if we want to do it it might be served better as a follow up step.

woutermont commented on proposals/specification/agent-management/ard-response.bs:
```diff
@@ -0,0 +1,62 @@
+
+#### Error Responses #### {#sec-ard-response-error}
+
+- If no [:Authorization:] header is present, the [=Agent Manager=] [MUST] respond with status code [S401], including a [:WWW-Authenticate:] header with the [:DPoP:] scheme. This response [MUST NOT] include an [:error:] parameter.
```
Not sure I understand. What would the Agent Manager use as who's Identity Document?

elf-pavlik replied:
In [sai-impl-service](https://github.com/janeirodigital/sai-impl-service/blob/main/packages/service/src/handlers/agents-handler.ts) the AuthorizationAgent needs to get data from the solid storage. It has privileged access set to the end user (agent) and the authrization agent (client). In that case the Authorization Agent acts as Solid OIDC client and serves its identity document from the same IRI.

woutermont replied:
Okay, but what is preventing that?

woutermont commented:
Ass requested by @elf-pavlik in yesterday's meeting, here's a quickly drafted snippet of how a client would discover its registration, with support for the "legacy" use of `pim:preferencesFile`, reading links out of the body rather than the headers, and supporting all reponse codes (`303`, `204` and `200`). I've commented out (with `//`) those parts that would be superfluous when we would not provide this additional support (only read a single relation, only out of headers, and with only `204` responses) ... it's not much difference i.m.o.

The only thing I left out is the `rdfs:seeOther` link, because I must agree with @elf-pavlik that this would put too much burden on the client. So we only check for one or two predicates/links of which we specify that they should be unique.

```
/* RDF Parser */
// import { Parser } from 'n3';
// const parser = new Parser();

/* Shorthands */
const hasAM = 'http://www.w3.org/ns/solid/interop#hasAgentManager';
const regAgent = 'http://www.w3.org/ns/solid/interop#registeredAgent';
// const prefFile = 'https://www.w3.org/ns/pim/space#preferencesFile';

/* Helper functions */
// const const rdfLinks = (rdf) => Object.fromEntries(parser.parse(rdf)
//   .map((quad) => [quad.predicate.value, { target: quad.object.value, context: quad.subject.value }]))
const headerLinks = (response) => Object.fromEntries(
  response.headers?.get('link')?.split(',')?.map((link) => link.trim().split('; ')).map((params) => [
    params.find((param) => param.startsWith('rel=')).replaceAll(/(^rel="|"$)/g,''), ({
      target: params[0].replaceAll(/(^<|>$)/g,''),
      context: params.find((param) => param.startsWith('anchor='))?.replaceAll(/(^anchor="|"$)/g,'') ?? response.url
    })
  ])
)

const getRegistration = async (webid) {

  /* Get links from profile header and body */
  const profile = await fetch(webid);
  const hlinks = headerLinks(profile)
  let am = hlinks[hasAM];
  // let pf = hlinks[prefFile];
  // if (!am && !pf) {
  //   const rlinks = rdfLinks(await profile.text());
  //   am = rlinks[hasAM];
  //   pf = rlinks[prefFile];
  // }

  /* Abort if no indication of agent manager  */
  if (!am && !pf) return undefined; 

  /* Request agent registration */
  const token = getToken(profile);
  const target = am?.target ?? pm?.target
  const dpop = getDPoP('GET', target);
  const registration = await fetch(target, {
    redirect: manual,
    headers: {
      Authorization: `DPoP ${token}`,
      DPoP: dpop,
    },
  });

  /* Optimal case, we're done */
  // if (registration.status === 200) return parser.parse(await registration.text());

  /* Otherwise, follow registration link, then we're done */
  if (registration.status === 204 || registration.status === 303) { 
    const reg = headerLinks(registration)[regAgent].context;
    // const loc = registration.headers.get('location');
    const dpop = getDPoP('GET', reg ?? loc);
    const registration = await fetch(reg ?? loc, {
      headers: {
        Authorization: `DPoP ${token}`,
        DPoP: dpop,
      },
    });
    return parser.parse(await registration.text());
  }

  /* All other cases indicate some auth error
   * which SAI-aware clients could have foreseen. */
  throw new Error('Auth error');

}
```

PS: I wrote this code directly into this comment, so please treat it as pseudocode; no idea if it would work without some finetuning.

woutermont commented on proposals/specification/agent-management/amd-response.bs:
```diff
@@ -0,0 +1,51 @@
+omitted, in which case the [=Identity Document=] [SHOULD] contain an [=RDF
+Triple=] linking the [=Entity's=] [=Identifier=] to the URL of its [=Agent
+Manager=] with the predicate `interop:hasAgentManager`. The predicate [MAY] also
+be substituted by `rdfs:seeAlso` or `pim:preferencesFile`.
```
ACTION: remove `rdfs:seeAlso`

TallTed commented on proposals/specification/agent-management/arr-request.bs:
```diff
@@ -0,0 +1,26 @@
+
+    </div>
+
+    ADVISEMENT: The exact implementation of the process that follows is up to the implementation of the [=Agent Manager=] itself. However, in case the [=Agent=] managing the [=Identity Document=] of the [=Entity=] is present, it is advisable that the process at least contains an interaction sequence between the [=Agent Manager=] and this managing [=Agent=]. 
```
```suggestion
    ADVISORY: The exact implementation of the process that follows is up to the implementation of the [=Agent Manager=] itself. However, if the [=Agent=] managing the [=Identity Document=] of the [=Entity=] is present, it is advisable that the process contains at least an interaction sequence between the [=Agent Manager=] and this managing [=Agent=]. 
```

TallTed commented on proposals/specification/agent-management/arr-response.bs:
```diff
@@ -0,0 +1,27 @@
+    
+    </div>
+
+    NOTE: Responses in case of invalid or denied requests are entirely left to the implementation.
```
```suggestion
    NOTE: Responses to invalid or denied requests are entirely left to the implementation.
```

TallTed commented on proposals/specification/agent-management/arr-result.bs:
```diff
@@ -0,0 +1,8 @@
+
+## Registration Result ## {#sec-arr-result}
+
+In case of a successful response, i.e. when the status code is [S201]
+or [S303], the [=Agent=] [MAY] retrieve the location of its [=Agent
```
```suggestion
If a response is successful (that is, if the status code is [S201]
or [S303]), the [=Agent=] [MAY] retrieve the location of its [=Agent
```

TallTed commented on proposals/specification/agent-management/considerations.bs:
```diff
@@ -0,0 +1,64 @@
+Some considerations can be made when conceptualizing the [=Agent Manager=], and
+its relation to other [=Resources=] and inquiring [=Agents=].
+
+1. In order to use a [=Agent Manager=] to discover the [=Agent-Specific
```
```suggestion
1. To use an [=Agent Manager=] to discover the [=Agent-Specific
```

TallTed commented on proposals/specification/agent-management/considerations.bs:
```diff
@@ -0,0 +1,64 @@
+    1. The response provided by the [=Agent Manager=] should be applicable to
+       all document types the other service might respond with.
+
+    2. The response provided by a [=Agent Manager=] should be minimal, as not to
```
```suggestion
    2. The response provided by an [=Agent Manager=] should be minimal, so as not to
```

TallTed commented on proposals/specification/agent-management/considerations.bs:
```diff
@@ -0,0 +1,64 @@
+       all document types the other service might respond with.
+
+    2. The response provided by a [=Agent Manager=] should be minimal, as not to
+       cause too much bloat in case the purpose of a request was purely the
```
```suggestion
       cause too much bloat if the purpose of a request was purely the
```

TallTed commented on proposals/specification/agent-management/considerations.bs:
```diff
@@ -0,0 +1,64 @@
+
+    4. While a read operation on the combined service might combine the info of
+       both services (in a distinguishable way), the presence of the [=Agent
+       Manager=] should not disturb the other operations on the other service
+       (e.g. write operations).
```
```suggestion
       Manager=] should not disturb other operations (such as write) on the
       other service.
```

TallTed commented on proposals/specification/agent-management/diagrams/agent-management.seq.mmd:
```diff
@@ -0,0 +1,37 @@
+    ID-->>App: 200 OK
+
+    break No Agent Manager <br/> in Link header or Body
+        Note over ID,App: Rely on Identity Document only
```
```suggestion
        Note over ID,App: Rely only on Identity Document
```

TallTed commented on proposals/specification/agent-management/diagrams/agent-management.seq.mmd:
```diff
@@ -0,0 +1,37 @@
+    AM-->>App: 200 OK
+
+    break No Agent Registration <br/> or Registration Endpoint<br/>in Link header or Body
+        Note over App: Rely on Identity Document only
```
```suggestion
        Note over App: Rely only on Identity Document
```

TallTed commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+Abstract: 
+
+  This specification details how Agents can discover the Resource Discovery 
+  Hub of an Entity, and how they can interact with it in order to find 
```
```suggestion
  Hub of an Entity, and how they can interact with it to find 
```

TallTed commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+  [contribute](https://github.com/solid/data-interoperability-panel/issues) any
+  feedback, comments, or questions you might have. Publication as an Editor's
+  Draft does not imply endorsement by the W3C Membership. This is a draft
+  document and may be updated, replaced or obsoleted by other documents at any
```
```suggestion
  document and may be updated, replaced, or obsoleted by other documents at any
```

TallTed commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+  feedback, comments, or questions you might have. Publication as an Editor's
+  Draft does not imply endorsement by the W3C Membership. This is a draft
+  document and may be updated, replaced or obsoleted by other documents at any
+  time. It is inappropriate to cite this document as other than work in
```
```suggestion
  time. It is inappropriate to cite this document as anything other than work in
```

TallTed commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+  progress. This document was produced by a group operating under the [W3C
+  Community Contributor License Agreement
+  (CLA)](https://www.w3.org/community/about/process/cla/). A human-readable
+  [summary](https://www.w3.org/community/about/process/cla-deed/) is available.
```
```suggestion
  [summary of the CLA](https://www.w3.org/community/about/process/cla-deed/) is available.
```

TallTed commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+<link rel="stylesheet" href="style.css" />
+
+<div boilerplate=copyright>
+MIT License Copyright 2019 - 2021 W3C Solid Community Group
```
```suggestion
MIT License Copyright 2019 - 2023 W3C Solid Community Group
```

TallTed commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+[NONORM]
+
+When [=Agents=] wish to **discover** information about an [=Entity=] 
+(e.g. a person, document or API), 
```
```suggestion
(such as a person, document, or API), 
```

TallTed commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+(e.g. a person, document or API), 
+they typically dereference a [=Dereferenceable Identifier=] of it, 
+to find a single [=Identity Document=], 
+e.g. the [=Profile Document=] of a [=WebID=], 
+or [=DID Document=] of a [=Decentralized Identifier=]. 
```
```suggestion
such as the [=Profile Document=] of a [=WebID=], 
or the [=DID Document=] of a [=Decentralized Identifier=]. 
```

TallTed commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+However, an [=Agent=] managing the [=Identity Document=] of the [=Entity=] 
+might want to expose *different* information about the [=Entity=], 
+depending on the [=Agent=] trying to discover it. 
+This <dfn>Managing Agent</dfn> might want this 
+to restrict access for privacy or security concerns,
+or to provide information specific to particular [=Agents=]: 
+tailored indexes, user preferences, application state etc.
```
```suggestion
This <dfn>Managing Agent</dfn> might want
to restrict access for privacy or security concerns,
or to provide information specific to particular [=Agents=],
like tailored indexes, user preferences, application state, etc.
```

TallTed commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+
+This document specifies a process by which [=Agents=] can discover 
+this limited set of <dfn export>Agent-Specific Resources</dfn> &mdash; 
+[=Resources=] that are accessible by and possibly relevant for a specific [=Agent=] 
```
```suggestion
[=Resources=] that are accessible by and possibly relevant to a specific [=Agent=] 
```

TallTed commented on proposals/specification/agent-management/terms.bs:
```diff
@@ -0,0 +1,163 @@
+This document builds on the concepts and definitions of [RDF]. 
+For conformance, terms from [[!RFC9110|HTTP Semantics]] and [LINKS] are used. 
+[[WEBID|WebID's]] and [[DID|Decentralized Identifiers]] are used as examples. 
+References to all these are linked throughout the text, 
```
```suggestion
References to all of these are linked throughout the text, 
```

TallTed commented on proposals/specification/agent-management/terms.bs:
```diff
@@ -0,0 +1,163 @@
+    <dfn>RDF-Dereferenceable</dfn> [=Identifiers=] dereference to an [=RDF Document=].
+
+: <dfn>Agent</dfn> 
+:: An actor, i.e. an [=Entity=] that acts, does stuff, or has the power to. 
```
```suggestion
:: An actor (that is, an [=Entity=]) that acts, does stuff, or has the power to do so. 
```

TallTed commented on proposals/specification/agent-management/terms.bs:
```diff
@@ -0,0 +1,163 @@
+
+: <dfn>Agent</dfn> 
+:: An actor, i.e. an [=Entity=] that acts, does stuff, or has the power to. 
+    It can do this on its own, or mediated by some other [=Agent=]. 
```
```suggestion
    It can do this on its own, or through mediation by some other [=Agent=]. 
```

TallTed commented on proposals/specification/data-authorization.bs:
```diff
@@ -185,27 +187,16 @@ are processed by the [=Authorization Agent=].
 ### Authorization Agent Discovery ### {#authorization-agent-discovery}
 
+This document extends the [Agent Manager Discovery](https://solid.github.io/data-interoperability-panel/specification/agent-management/index.html#sec-am-discovery) of the [[sai-am|Agent Manager]] specification,
+by defining `interop:hasAuthorizationAgent` as subproperty of `interop:hasAccessManager`.
```
```suggestion
by defining `interop:hasAuthorizationAgent` as a subproperty of `interop:hasAccessManager`.
```

TallTed commented on proposals/specification/data-authorization.bs:
```diff
@@ -185,27 +187,16 @@ are processed by the [=Authorization Agent=].
 
+This document extends the [Agent Manager Discovery](https://solid.github.io/data-interoperability-panel/specification/agent-management/index.html#sec-am-discovery) of the [[sai-am|Agent Manager]] specification,
+by defining `interop:hasAuthorizationAgent` as subproperty of `interop:hasAccessManager`.
+the [=Authorization Agent=] for a given [=Social Agent=] can be discovered
```
```suggestion
```

justinwb reviewed (changes requested):
Overall looks pretty good - a few comments / questions

TallTed commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+This document specifies a process by which [=Agents=] can discover 
+this limited set of <dfn export>Agent-Specific Resources</dfn> &mdash; 
+[=Resources=] that are accessible by and possibly relevant for a specific [=Agent=] 
+&mdash; in a performant, secure, and privacy-minded manner, 
+without having to bother with all other, irrelevant [=Resources=].
```
```suggestion
&mdash; in a performant, secure, and privacy-minded manner.
```

TallTed replied:
or possibly
```suggestion
&mdash; in a performant, secure, and privacy-minded manner, 
without trying (and failing) to access all other, irrelevant [=Resources=].
```

woutermont commented on proposals/specification/agent-management/index.bs:
```diff
@@ -0,0 +1,139 @@
+Abstract: 
+
+  This specification details how Agents can discover the Resource Discovery 
+  Hub of an Entity, and how they can interact with it in order to find 
```
Correct, should be _Agent Manager_ now.

woutermont commented on proposals/specification/agent-management/intro.bs:
```diff
@@ -0,0 +1,31 @@
+this limited set of <dfn export>Agent-Specific Resources</dfn> &mdash; 
+[=Resources=] that are accessible by and possibly relevant for a specific [=Agent=] 
+&mdash; in a performant, secure, and privacy-minded manner, 
+without having to bother with all other, irrelevant [=Resources=].
```
It stresses the _efficiency gain_ of using an Agent Manager, instead of needing to traverse all Resources in order to find out which ones are accessible and relevant.

Do you think we should leave it out?

woutermont commented:
Closing this since I am no longer convinced that a separate discovery service is needed (cf. https://github.com/solid/data-interoperability-panel/issues/315). An agent should at all times be able to request access to (types of) resources at an AS, and the combination of the AS's response and a token info endpoint should suffice to gain all necessary information. If not, a better approach than proposed in this PR would be to add a client/agent info endpoint that functions similarly to the OIDC user info endpoint.
